text/0000-try-clone-trait.md

```diff
@@ -0,0 +1,72 @@
+- try-clone-trait
+- Start Date: 2017-10-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+The standard library provides the `Clone` trait to duplicate an object. However,
+there is no way to indicate fallibility of the operation. This RFC proposes the
+`TryClone` trait that supports fallible clones. 
+
+# Motivation
+Failing to clone an object is fairly common, for example, the system runs out of resources to represent the underlying type.
+Implementations have worked around this in two ways. Some types in the standard library has a `try_clone` method that returns a 
+`Result<T>`. A number of [crates](https://github.com/search?l=Rust&q=TryClone&type=Code&utf8=%E2%9C%93) have defined `TryClone` as a trait and implemented it on standard
+types.
+
+One example is [Clone](https://doc.rust-lang.org/src/alloc/arc.rs.html#710) for `std::alloc::arc`. This is better represented
+by returning a `Result<Arc>`. Doing that will also make the unsafe block unnecessary. Also, this will enable
+the caller to handle these errors idiomatically instead of aborting the process.
```

Comment:
This error is only encountered if your process leaks 2 billion copies of the Arc, after which you can never clone it again. How is that situation supposed to be "idiomatically handled"?

Reply:
Talking about 2 billion copies... An `Arc` takes 4 bytes on 32-bit architectures plus the data behind that 4 bytes pointer. If, in theory, you could use every single byte of your program to allocate a copy of an `Arc`, then you would have a maximum reference count of `0x40000000_u32`, which is `< u32::max_value()`. And now reduce that number further, because you also have to fit the actual data, your whole program, and the OS kernel into your 4GiB address space.

As a result, it is absolutely impossible to overflow the reference counters of `Arc` and `Rc`, ever, on any CPU architecture. So, in a way, you could optimise `alloc` instead by not doing overflow checks.

Reply:
```rust
let foo = Arc::new(0);
loop {
    mem::forget(foo.clone());
}
```

Reply:
Now, that is just evil. ;)